src/primitives.rs

```diff
@@ -0,0 +1,117 @@
+//! Contains comparison primitives used to build up the rest of the library.
+
+/// Returns list of triples describing matching sequences.
+///
+/// The first number is the index in the first string of the beginning of the match.
+/// The second number is the index of the second string of the beginning of the match.
+/// The final number is the length of the match.
+///
+/// The final matching sequence will be a trivial matching sequence of (a.len(),
+/// b.len(), 0) and will be the only match of length 0.
+///
+/// ```
+/// # use fuzzywuzzy::segmentation::{Segmenter, CodePointSegmenter, GraphemeSegmenter};
+/// # use fuzzywuzzy::primitives::get_matching_blocks;
+/// assert_eq!(get_matching_blocks(&CodePointSegmenter.segment("abxcd"), &CodePointSegmenter.segment("abcd")), vec![(0, 0, 2), (3, 2, 2), (5, 4, 0)]);
+/// assert_eq!(get_matching_blocks(&CodePointSegmenter.segment("abcd"), &CodePointSegmenter.segment("abxcd")), vec![(0, 0, 2), (2, 3, 2), (4, 5, 0)]);
+/// assert_eq!(get_matching_blocks(&CodePointSegmenter.segment("chance"), &CodePointSegmenter.segment("スマホでchance")), vec![(0, 4, 6), (6, 10, 0)]);
+/// assert_eq!(get_matching_blocks(&GraphemeSegmenter.segment("chance"), &GraphemeSegmenter.segment("スマホでchance")), vec![(0, 4, 6), (6, 10, 0)]);
+/// assert_eq!(get_matching_blocks(&CodePointSegmenter.segment("किमप"), &CodePointSegmenter.segment("किमपि")), vec![(0, 0, 4), (4, 5, 0)]);
+/// assert_eq!(get_matching_blocks(&GraphemeSegmenter.segment("किमप"), &GraphemeSegmenter.segment("किमपि")), vec![(0, 0, 2), (3, 3, 0)]);
+/// ```
+#[allow(clippy::many_single_char_names)]
+pub fn get_matching_blocks<T: Eq>(a: &[T], b: &[T]) -> Vec<(usize, usize, usize)> {
+    let flipped;
+    let (shorter, len1, longer, len2) = {
+        let a_len = a.len();
+        let b_len = b.len();
+        if a_len <= b_len {
+            flipped = false;
+            (a, a_len, b, b_len)
+        } else {
+            flipped = true;
+            (b, b_len, a, a_len)
+        }
+    };
+    // https://github.com/python-git/python/blob/master/Lib/difflib.py#L461
```

Comment:
I don't have time to review and align the implementation details from this Python source, but I trust translating any relevant unit-tests for this logic will be sufficient.

Reply:
I translated this from the impl in util that only worked on `&str`. but adding canonical tests from the source would be good.

Reply:
Agreed, just having a simple guarantee that the behavior is unchanged from the original source would be preferable.

Reply:
testing is extremely implicit for this function in cpython

https://github.com/python/cpython/blob/63298930fb531ba2bb4f23bc3b915dbf1e17e9e1/Lib/test/test_difflib.py